plugins/imjournal/imjournal.c

```diff
@@ -431,83 +440,70 @@ persistJournalState(void)
 	DEFiRet;
 	FILE *sf; /* state file */
 	char tmp_sf[MAXFNAME];
-	char *cursor;
-	int ret = 0;
-
-	/* On success, sd_journal_get_cursor() returns 1 in systemd
-	   197 or older and 0 in systemd 198 or newer */
-	if ((ret = sd_journal_get_cursor(j, &cursor)) >= 0) {
-               /* we create a temporary name by adding a ".tmp"
-                * suffix to the end of our state file's name
-                */
-               snprintf(tmp_sf, sizeof(tmp_sf), "%s.tmp", cs.stateFile);
-               if ((sf = fopen(tmp_sf, "wb")) != NULL) {
-			if (fprintf(sf, "%s", cursor) < 0) {
-				iRet = RS_RET_IO_ERROR;
-			}
-			fclose(sf);
-			free(cursor);
-                       /* change the name of the file to the configured one */
-                       if (iRet == RS_RET_OK && rename(tmp_sf, cs.stateFile) == -1) {
-                               LogError(errno, iRet, "imjournal: rename() failed: "
-                                       "for new path: '%s'", cs.stateFile);
-                               iRet = RS_RET_IO_ERROR;
-                       }
 
-		} else {
-			LogError(errno, RS_RET_FOPEN_FAILURE, "imjournal: fopen() failed "
-				"for path: '%s'", tmp_sf);
-			iRet = RS_RET_FOPEN_FAILURE;
-		}
-	} else {
-		LogError(-ret, RS_RET_ERR, "imjournal: sd_journal_get_cursor() failed");
-		iRet = RS_RET_ERR;
+	/* first check that we have valid cursor */
+	if (!last_cursor) {
+		ABORT_FINALIZE(RS_RET_OK);
 	}
+
+	/* we create a temporary name by adding a ".tmp"
+	 * suffix to the end of our state file's name
+	 */
+	snprintf(tmp_sf, sizeof(tmp_sf), "%s.tmp", cs.stateFile);
+
+	sf = fopen(tmp_sf, "wb");
+	if (sf == NULL) {
+		LogError(errno, RS_RET_FOPEN_FAILURE, "imjournal: fopen() failed for path: '%s'", tmp_sf);
+		ABORT_FINALIZE(RS_RET_FOPEN_FAILURE);
+	}
+
+	iRet = fputs(last_cursor, sf);
```

Comment:
We need an `fsync()` here as well.

Reply:
Won't the `fclose()` do a sync?

Reply:
> Won't the fclose() do a sync?

Yes, it will, according to man fclose:
```
       The  fclose() function flushes the stream pointed to by stream (writing
       any buffered output data using fflush(3))  and  closes  the  underlying
       file descriptor.
```

Reply:
See https://lwn.net/Articles/457667/

Reply:
Yes, particularity [this](https://lwn.net/Articles/458176/ ) comment, I do not really like the idea of such a load here.